### PR TITLE
net: lwm2m: Bootstrap procedure improvements

### PR DESCRIPTION
**Pull request by rlubos**
This PR introduces several improvements and fixes for LwM2M bootstrap procedure. Detailed description of the changes can be found in the commit message, to summarize:

Improvements:
 * Introduced new state machine state, `ENGINE_IDLE`, which now is the default after boot and after LwM2M engine is stopped.
 * Allow to specify whether bootstrap procedure shall run in the current session, instead of relying on static configuration only.

Fixes:
 * Introduce new member in the resource instance structure, indicating actual data size stored (so far only the buffer size was kept, making it impossible to retrieve the actual opaque data size written by the bootstrap server).
 * Fix PSK ID provisioning, by using the actual data size instead of result of `strlen()` (the PSK ID written by the server is of opaque data type and is not NULL terminated).
 * Report Bootstrap Complete event only after all bootstrap transactions are done.

Based on https://github.com/zephyrproject-rtos/zephyr/pull/28592, marking DNM until it's merged.

### CODE REVIEW

**Review by jukkar — APPROVED** [review]
LGTM

File: include/net/lwm2m.h [hunk 1]
```diff
@@ -871,12 +871,13 @@ typedef void (*lwm2m_ctx_event_cb_t)(struct lwm2m_ctx *ctx,
...
  * @param[in] client_ctx LwM2M context
  * @param[in] ep_name Registered endpoint name
+ * @param[in] bootstrap Run bootstrap procedure in this session.
  * @param[in] event_cb Client event callback function
  *
  * @return 0 for success or negative in case of error.
  */
 void lwm2m_rd_client_start(struct lwm2m_ctx *client_ctx, const char *ep_name,
-			   lwm2m_ctx_event_cb_t event_cb);
+			   bool bootstrap, lwm2m_ctx_event_cb_t event_cb);
```
Comment (carlescufi) [inline include/net/lwm2m.h:880]: Let's use an `flags` integer instead so we are future-proof.

**Review by carlescufi — CHANGES_REQUESTED** [review]

File: include/net/lwm2m.h (hunk 1 above)
Comment (rlubos) [reply include/net/lwm2m.h]: I've added `flags` parameter as suggested and a release notes entry.

**Review by carlescufi — APPROVED** [review]

File: include/net/lwm2m.h [hunk 2]
```diff
@@ -871,12 +880,13 @@ typedef void (*lwm2m_ctx_event_cb_t)(struct lwm2m_ctx *ctx,
  *
  * @param[in] client_ctx LwM2M context
  * @param[in] ep_name Registered endpoint name
+ * @param[in] flags Flags used to configure current LwM2M session.
  * @param[in] event_cb Client event callback function
  *
  * @return 0 for success or negative in case of error.
  */
 void lwm2m_rd_client_start(struct lwm2m_ctx *client_ctx, const char *ep_name,
```
Comment (jukkar) [inline include/net/lwm2m.h:888]: Just noticed this, the function is void but comment says it returns 0 for success and <0 for error. Which one is correct?

Comment (rlubos) [reply include/net/lwm2m.h:888]: It's void, I'll remove the invalid docs.

**Comment by rlubos** [comment]
@jukkar I've removed the invalid API doc entries.

@carlescufi I've added separate entries for application protocol APIs (CoAP, MQTT, LwM2M) to the overview page, as requested. Not sure if there's any pattern in adding these entries (they're not sorted alphabetically), for now I've added the new entries just below "Networking".

**Review by carlescufi — APPROVED** [review]

**Pull request merged**